インクリメンタルサーチの実装

SK2929 opened the pull request:
# what
インクリメンタルサーチを実装

# why
ユーザー検索をインクリメンタルサーチ化するため

Commit:
first commit

Commit:
Create-user.js

2. ルーティングの設定

Commit:
Add-add_index_to_users.rb

Usersテーブルのnameカラムを
インデックスでデータの検索を高速化

Commit:
Edit-_form.html.haml

3. テキストフィールドを作成する

Commit:
Edit-user.js

4. テキストフィールドに入力されるたびにイベントが発火するようにする

Commit:
Edit-jbuilder

5. イベント時に非同期通信できるようにする
2.で作成したルーディングに対する
・users_controller.rbでrespond_toを使ったAPI記述

Commit:
Edit-index.json.jbuilder

5. イベント時に非同期通信できるようにする
2.で作成したルーディングに対する
・Jbuilderの変数定義

Commit:
Edit-user.js

6. 非同期通信の結果を得て、HTMLを作成する（4行目〜17行目）

7. 6で作成したHTMLをビュー上に追加する（2行目）

8. エラー時の処理を行う
（35行目〜37行目）

hzak2xx commented on app/assets/javascripts/user.js:
```diff
@@ -0,0 +1,39 @@
+function appendProduct(user) {
+  var html = `<div class="chat-group-user clearfix">
+                <p class="chat-group-user__name">${ user.name }</p>
+                <a class="user-search-add chat-group-user__btn chat-group-user__btn--add" data-user-id="ユーザーのid" data-user-name="${ user.name }">追加</a>
```
`ユーザーのid` 修正しましょう

SK2929 replied:
ユーザーのidを変数で取得できるように変更
それに伴って、index.json.jbuilderの変数項目を追加しました

hzak2xx commented on app/controllers/users_controller.rb:
```diff
@@ -1,6 +1,11 @@
 class UsersController < ApplicationController
 
 def index
+  @users = User.where('name LIKE(?)', "%#{params[:keyword]}%")
```
current_user取得しないようにしましょう

SK2929 replied:
Where.notを使用して、ログイン済みのユーザーは表示しないように変更しました

hzak2xx commented on app/views/groups/_form.html.haml:
```diff
@@ -14,29 +14,26 @@
-        </div>
+      #chat-group-users
+        #chat-group-user-22.chat-group-user.clearfix
+          %input{name: "chat_group[user_ids][]", type: "hidden", value: "22"}/
```
`22`修正しましょう

SK2929 replied:
22という記述を削除しました

Commit:
Edit-user.js、index.json.jbuilder

修正事項：ユーザーのid 修正しましょう

ユーザーのidを変数で取得できるように変更
それに伴って、
index.json.jbuilderの変数項目を追加

Commit:
Edit-users_controller.rb

指摘事項：current_user取得しないようにしましょう

Where.notを使用して、ログイン済みのユーザーは表示しないように変更

Commit:
Edit-_form.html.haml

指摘事項：22修正しましょう

22という記述を削除

ky3090 commented on app/views/groups/_form.html.haml:
```diff
@@ -14,29 +14,26 @@
+      #chat-group-users
+        #chat-group-user-.chat-group-user.clearfix
+          %input{name: "chat_group[user_ids][]", type: "hidden", value: ""}/
+          %p.chat-group-user__name seo_kyohei
```
> %input{name: "chat_group[user_ids][]", type: "hidden", value: ""}/

hidden_fieldを使いましょう。
seo_kyoheiではなく、userを表示させましょう。
そしてvalueに値を取れるようにしましょう

SK2929 replied:
１、hidden_fieldで表現しました
２、=@current_user.nameで表現しました
３、value: group.idと追加しました

修正後の記述
          =f.hidden_field :id, name: "chat_group[user_ids][]", value: group.id
          %p.chat-group-user__name
            = @current_user.name

ky3090 commented on app/controllers/users_controller.rb:
```diff
@@ -1,6 +1,11 @@
 class UsersController < ApplicationController
 
 def index
+  @users = User.where('name LIKE(?)', "%#{params[:keyword]}%").where.not(id: current_user.id)
```
> where.not(id: current_user.id)

where()だけで済む方法を考えてみましょう。

ヒントはnot equal

SK2929 replied:
.where('name LIKE(?) AND id != ?', "%#{params[:keyword]}%", current_user)
という記述に変更しました

ky3090 commented on app/views/groups/_form.html.haml:
```diff
@@ -14,29 +14,26 @@
-      = f.collection_check_boxes :user_ids, User.all, :id, :name
+      / = f.collection_check_boxes :user_ids, User.all, :id, :name
       / この部分はインクリメンタルサーチ（ユーザー追加の非同期化のときに使用します
       /
```
必要のないコメントアウトは削除しましょう！

SK2929 replied:
削除しました

ky3090 commented on app/views/groups/_form.html.haml:
```diff
@@ -14,29 +14,26 @@
   .chat-group-form__field.clearfix
     / この部分はインクリメンタルサーチ（ユーザー追加の非同期化のときに使用します
     /
```
ここもです。

SK2929 replied:
削除しました

ky3090 reviewed (commented):
挙動のgif画像も一緒にください！

SK2929 commented:
![5cf5e872083126621553097c7553e4af](https://user-images.githubusercontent.com/44630490/49207827-69121880-f3f9-11e8-8c21-74906696270b.gif)
hirozakさんの指摘事項を修正後の挙動のGifです【ログイン中のユーザはSKです】

Commit:
Edit-users_controller.rb

指摘事項：where()だけで済む方法を考えてみましょう。

.where.not()表現を.where()にまとめた

Commit:
Edit-_form.html.haml

指摘事項：
１、hidden_fieldを使いましょう。
２、seo_kyoheiではなく、userを表示させましょう。
３、valueに値を取れるようにしましょう

１、hidden_fieldで表現
２、= @current_user.nameで表現
３、value: group.idと追加

SK2929 commented:
![5cf5e872083126621553097c7553e4af](https://user-images.githubusercontent.com/44630490/49222034-20208b00-f41e-11e8-8f39-ca0821fd2908.gif)
yutakatahiraさんの指摘事項修正後のGif

SK2929 commented:
質問：Edit-_form.html.hamlの29行目の記述が、現在下記の状態なのですが
#chat-group-user-.chat-group-user.clearfix
このままで大丈夫でしょうか？
